tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Session.java

```diff
@@ -135,6 +143,111 @@ public void close() {
     }
   }
 
+  /**
+   * The result of a run in a session.  Contains the fetched tensors and the outputs that were fetched.
+   * <p>
+   * Closing a {@code Result} object will close all of the tensors contained by it.
+   */
+  public final class Result implements AutoCloseable, Iterable<Tensor<?>>{
+    private final List<Tensor<?>> results;
+    private final List<Output<?>> fetches;
+    private final LinkedHashMap<Output<?>, Integer> indexMap;
+
+    private Result(List<Tensor<?>> results, List<Output<?>> fetches) {
+      this.results = new ArrayList<>(results);
+      this.fetches = new ArrayList<>(fetches);
+      indexMap = new LinkedHashMap<>();
+      for(int i = 0 ; i < fetches.size() ; i++){
+        indexMap.put(fetches.get(i), i);
+      }
+    }
+
+    /**
+     * Get the result tensors.
+     */
+    public List<Tensor<?>> getResults() {
+      return results;
+    }
+
+    /**
+     * Get the outputs that were fetched.
+     */
+    public List<Output<?>> getFetches() {
+      return fetches;
+    }
+
+    /**
+     * Get the result at {@code index}.
+     */
+    public Tensor<?> get(int index){
+      return results.get(index);
+    }
+
+    /**
+     * Get the result for {@code output} or throw an {@code IllegalArgumentException} if it wasn't fetched.
```

Comment:
Given these get methods are basically Map lookups it's a bit unfriendly to have them throw rather than return null or `Optional<>` as would be idiomatic for Java.

Reply:
I'm of two minds here.  On one hand, it is a map lookup, so that makes sense.  On the other hand, since everything in this map is stuff you passed to `fetch`, you will most likely only be calling `get` with things you know exist, and any times you don't will most likely be errors for which throwing makes more sense.  I could always add `getOrError` but that seems like overkill especially w/ the added contains methods.  Thoughts?

Reply:
Also, now that I have a getter for the map, if you want `null` returns you can just use that.

Reply:
@Craigacp thoughts on renaming it to `fetch` (w/ the same behavior as now)?